### ISSUE TITLE
Building a Better Profanity Detection Library with scikit-learn - victorzhou.com

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2870]
Building a Better Profanity Detection Library with scikit-learn - victorzhou.com<br>
<br>
https://ift.tt/2SOHslL<br>
<br>
Victor Zhou<br>
<br>
<div><div>
<p>A few months ago, I needed a way to detect profanity in user-submitted text strings:</p> <p><figure><img src="https://cdn-images-1.medium.com/max/1600/1*i2fk4aGvplR7le_3PPajAA.png"><figcaption>This shouldn’t be that hard, right?</figcaption></figure></p> <p>I ended up building and releasing my own library for this purpose called <a rel="noopener noreferrer" href="https://github.com/vzhou842/profanity-check" target="_blank">profanity-check</a>.</p> <p>Of course, before I did that, I looked in the <a rel="noopener noreferrer" href="https://pypi.org/" target="_blank">Python Package Index</a> (PyPI) for any existing libraries that could do this for me. The only half decent results for the search query “profanity” were:</p> <p>Third-party libraries can sometimes be sketchy, though, so I did my due diligence on these 4 results.</p> <h2>
<a href="https://victorzhou.com/blog/better-profanity-detection-with-scikit-learn/#profanity-better-profanity-and-profanityfilter"></a>profanity, better-profanity, and profanityfilter</h2> <p>After a quick dig through the <code>profanity</code> repository, I found a file named <a rel="noopener noreferrer" href="https://github.com/ben174/profanity/blob/master/profanity/data/wordlist.txt" target="_blank">wordlist.txt</a>:</p> <p><figure><img src="https://cdn-images-1.medium.com/max/2000/1*0lTbmHR5WE7HZ8wCvLpqtg.png"><figcaption>NSFW</figcaption></figure></p> <p>The entire <code>profanity</code> library is just a wrapper over this list of 32 words! <code>profanity</code> detects profanity simply by looking for one of these words.</p> <p>To my dismay, <code>better-profanity</code> and <code>profanityfilter</code> both took the same approach:</p> <p>This is bad because <strong>profanity detection libraries based on wordlists are extremely subjective.</strong> For example, <code>better-profanity</code>’s wordlist includes the word “suck.” Are you willing to say that any sentence containing the word “suck” is profane? Furthermore, any hard-coded list of bad words will inevitably be incomplete — do you think <code>profanity</code>’s 32 bad words are the only ones out there?</p> <p><img src="https://victorzhou.com/media/xkcd-290.png"></p> <figcaption> Fucking Blue Shells. source: <a href="https://xkcd.com/290/" target="_blank">xkcd</a> </figcaption><p>Having already ruled out 3 libraries, I put my hopes on the 4th and final one: <code>profanity-filter</code>.</p> <h2>
<a href="https://victorzhou.com/blog/better-profanity-detection-with-scikit-learn/#profanity-filter"></a>profanity-filter</h2> <p><code>profanity-filter</code> uses Machine Learning! Sweet!</p> <p>Turns out, it’s <strong><em>really</em></strong> slow. Here’s a benchmark I ran in December 2018 comparing (1) <code>profanity-filter</code>, (2) my library <code>profanity-check</code>, and (3) <code>profanity</code> (the one with the list of 32 words):</p> <p><figure><img src="https://cdn-images-1.medium.com/max/1600/1*KRJEl4YHfSTk9PmmScIcUA.png"><figcaption>A human could probably do this faster than profanity-filter can</figcaption></figure></p> <p>I needed to be able to perform many predictions in real time, and <code>profanity-filter</code> was not even close to being fast enough. But hey, maybe this is a classic tradeoff of accuracy for speed, right?</p> <p>Nope.</p> <p><figure><img src="https://cdn-images-1.medium.com/max/1600/1*LYOeGE6vTXTAKhJ_W1fZgQ.png"><figcaption>At least profanity-filter is not dead last this time</figcaption></figure></p> <p>None of the libraries I’d found on PyPI met my needs, so I built my own.</p> <h2>
<a href="https://victorzhou.com/blog/better-profanity-detection-with-scikit-learn/#building-profanity-check-part-1-data"></a>Building profanity-check, Part 1: Data</h2> <p>I knew that I wanted <code>profanity-check</code> to base its classifications on data to avoid being subjective <em>(read: to be able to say I used Machine Learning)</em>. I put together a combined dataset from two publicly-available sources:</p> <p>Each of these datasets contains text samples hand-labeled by humans through crowdsourcing sites like <a rel="noopener noreferrer" href="https://www.figure-eight.com/" target="_blank">Figure Eight</a>.</p> <p>Here’s what my dataset ended up looking like:</p> <p><figure><img src="https://cdn-images-1.medium.com/max/1600/1*Bw_we8cbs-WOpWXOCxzSTg.png"><figcaption>Combined = Tweets + Wikipedia</figcaption></figure></p> <blockquote> <p>The Twitter dataset has a column named <code>class</code> that’s 0 if the tweet contains hate speech, 1 if it contains offensive language, and 2 if it contains neither. I classified any tweet with a <code>class</code> of 2 as “Not Offensive” and all other tweets as “Offensive.”</p> </blockquote> <blockquote> <p>The Wikipedia dataset has several binary columns (e.g. <code>toxic</code> or <code>threat</code>) that represent whether or not that text contains that type of toxicity. I classified any text that contained <em>any</em> of the types of toxicity as “Offensive” and all other texts as “Not Offensive.”</p> </blockquote> <h2>
<a href="https://victorzhou.com/blog/better-profanity-detection-with-scikit-learn/#building-profanity-check-part-2-training"></a>Building profanity-check, Part 2: Training</h2> <p>Now armed with a cleaned, combined dataset (which you can <a rel="noopener noreferrer" href="https://github.com/vzhou842/profanity-check/blob/master/profanity_check/data/clean_data.csv" target="_blank">download here</a>), I was ready to train the model!</p> <blockquote> <p>I’m skipping over how I cleaned the dataset because, honestly, it’s pretty boring— if you’re interested in learning more about preprocessing text datasets check out <a rel="noopener noreferrer" href="https://machinelearningmastery.com/clean-text-machine-learning-python/" target="_blank">this article</a> or <a rel="noopener noreferrer" href="https://medium.com/@datamonsters/text-preprocessing-in-python-steps-tools-and-examples-bf025f872908" target="_blank">this post</a>.</p> </blockquote> <div><pre><code><span>import</span> pandas <span>as</span> pd <span>from</span> sklearn<span>.</span>feature_extraction<span>.</span>text <span>import</span> CountVectorizer <span>from</span> sklearn<span>.</span>calibration <span>import</span> CalibratedClassifierCV <span>from</span> sklearn<span>.</span>svm <span>import</span> LinearSVC <span>from</span> sklearn<span>.</span>externals <span>import</span> joblib <span># Read in data</span> data <span>=</span> pd<span>.</span>read_csv<span>(</span><span>'clean_data.csv'</span><span>)</span> texts <span>=</span> data<span>[</span><span>'text'</span><span>]</span><span>.</span>astype<span>(</span><span>str</span><span>)</span> y <span>=</span> data<span>[</span><span>'is_offensive'</span><span>]</span> <span># Vectorize the text</span> vectorizer <span>=</span> CountVectorizer<span>(</span>stop_words<span>=</span><span>'english'</span><span>,</span> min_df<span>=</span><span>0.0001</span><span>)</span> X <span>=</span> vectorizer<span>.</span>fit_transform<span>(</span>texts<span>)</span> <span># Train the model</span> model <span>=</span> LinearSVC<span>(</span>class_weight<span>=</span><span>"balanced"</span><span>,</span> dual<span>=</span><span>False</span><span>,</span> tol<span>=</span><span>1e</span><span>-</span><span>2</span><span>,</span> max_iter<span>=</span><span>1e5</span><span>)</span> cclf <span>=</span> CalibratedClassifierCV<span>(</span>base_estimator<span>=</span>model<span>)</span> cclf<span>.</span>fit<span>(</span>X<span>,</span> y<span>)</span> <span># Save the model</span> joblib<span>.</span>dump<span>(</span>vectorizer<span>,</span> <span>'vectorizer.joblib'</span><span>)</span> joblib<span>.</span>dump<span>(</span>cclf<span>,</span> <span>'model.joblib'</span><span>)</span></code></pre></div> <figcaption> Are you also surprised the code is so short? Apparently <a href="https://scikit-learn.org/" target="_blank">scikit-learn</a> does everything. </figcaption><p>Two major steps are happening here: (1) vectorization and (2) training.</p> <h3>
<a href="https://victorzhou.com/blog/better-profanity-detection-with-scikit-learn/#vectorization-bag-of-words"></a>Vectorization: Bag of Words</h3> <p>I used <code>scikit-learn</code>’s <a rel="noopener noreferrer" href="https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html" target="_blank">CountVectorizer</a> class, which basically turns any text string into a vector by counting how many times each given word appears. This is known as a <a rel="noopener noreferrer" href="https://en.wikipedia.org/wiki/Bag-of-words_model" target="_blank">Bag of Words</a> (BOW) representation. For example, if the only words in the English language were <code>the</code>, <code>cat</code>, <code>sat</code>, and <code>hat</code>, a possible vectorization of the sentence <code>the cat sat in the hat</code> might be:</p> <p><figure><img src="https://cdn-images-1.medium.com/max/1600/1*sbnts1u_QFB_V-X5DSC3pg.png"><figcaption>“the cat sat in the hat” -&gt; [2, 1, 1, 1, 1]</figcaption></figure></p> <p>The <code>???</code> represents any unknown word, which for this sentence is <code>in</code>. Any sentence can be represented in this way as counts of <code>the</code>, <code>cat</code>, <code>sat</code>, <code>hat</code>, and <code>???</code>!</p> <p><figure><img src="https://cdn-images-1.medium.com/max/1600/1*-wONWZDab2gNQP3Rfdpt_A.png"><figcaption>A handy reference table for the next time you need to vectorize “cat cat cat cat cat”</figcaption></figure></p> <p>Of course, there are far more words in the English language, so in the code above I use the <code>fit_transform<span>(</span><span>)</span></code> method, which does 2 things:</p> <ul>
<li>
<strong>Fit:</strong> learns a vocabulary by looking at all words that appear in the dataset.</li> <li>
<strong>Transform</strong>: turns each text string in the dataset into its vector form.</li> </ul>
<h3>
<a href="https://victorzhou.com/blog/better-profanity-detection-with-scikit-learn/#training-linear-svm"></a>Training: Linear SVM</h3> <p>The model I decided to use was a Linear Support Vector Machine (SVM), which is implemented by <code>scikit-learn</code>’s <a rel="noopener noreferrer" href="https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html" target="_blank">LinearSVC</a> class. <a rel="noopener noreferrer" href="https://medium.com/machine-learning-101/chapter-2-svm-support-vector-machine-theory-f0812effc72" target="_blank">This post</a> and <a rel="noopener noreferrer" href="https://www.svm-tutorial.com/2014/11/svm-understanding-math-part-1/" target="_blank">this tutorial</a> are good introductions if you don’t know what SVMs are.</p> <blockquote> <p>The <a rel="noopener noreferrer" href="https://scikit-learn.org/stable/modules/generated/sklearn.calibration.CalibratedClassifierCV.html" target="_blank">CalibratedClassifierCV</a> in the code above exists as a wrapper to give me the <code>predict_proba<span>(</span><span>)</span></code> method, which returns a probability for each class instead of just a classification. You can pretty much just ignore it if that last sentence made no sense to you, though.</p> </blockquote> <p>Here’s one (simplified) way you could think about why the Linear SVM works: during the training process, the model learns which words are “bad” and how “bad” they are because those words appear more often in offensive texts. <strong>It’s as if the training process is picking out the “bad” words for me</strong>, which is much better than using a wordlist I write myself!</p> <p>A Linear SVM combines the best aspects of the other profanity detection libraries I found: it’s fast enough to run in real-time yet robust enough to handle many different kinds of profanity.</p> <h2>
<a href="https://victorzhou.com/blog/better-profanity-detection-with-scikit-learn/#caveats"></a>Caveats</h2> <p>That being said, <code>profanity-check</code> is far from perfect. Let me be clear: take predictions from <code>profanity-check</code> with a grain of salt because <strong>it makes mistakes.</strong> For example, its not good at picking up less common variants of profanities like “f4ck you” or “you b1tch” because they don’t appear often enough in the training data. You’ll never be able to detect <em>all</em> profanity (people will come up with new ways to evade filters), but <code>profanity-check</code> does a good job at finding most.</p> <h2>
<a href="https://victorzhou.com/blog/better-profanity-detection-with-scikit-learn/#profanity-check"></a>profanity-check</h2> <p><code>profanity-check</code> is open source and available on PyPI! To use it, simply</p> <div><pre><code>$ pip install profanity-check</code></pre></div> <p>How could <code>profanity-check</code> be even better? Feel free to reach out or comment with any thoughts or suggestions!</p> <hr>
<p><em>This article was originally posted on <a rel="noopener noreferrer" href="https://medium.com/@victorczhou/building-a-better-profanity-detection-library-with-scikit-learn-3638b2f2c4c2" target="_blank">Medium</a>.</em></p>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via  <br>
<br>
March 7, 2019 at 07:33PM
